[Investigations] - Unskip and refactor discover state tests

michaelolo24 opened the pull request:
Dependent on: https://github.com/elastic/kibana/pull/173015

- fixes https://github.com/elastic/kibana/issues/165663
- fixes https://github.com/elastic/kibana/issues/165747

[Flaky Tests (100/100)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4870)

<!--ONMERGE {"backportTargets":["8.12"]} ONMERGE-->

michaelolo24 committed:
fix discover_state file and unskip

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/esql/discover_timeline_state_integration.cy.ts:
```diff
@@ -33,7 +33,9 @@ import { login } from '../../../../tasks/login';
   addNameToTimelineAndSave,
+  confirmLeaveUnsavedTimeline,
   createNewTimeline,
+  createTimelineOptionsPopoverBottomBar,
```
this method does not exist in this file

michaelolo24 replied:
Rebase error, fixing

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/esql/discover_timeline_state_integration.cy.ts:
```diff
@@ -50,25 +52,29 @@ const INITIAL_START_DATE = 'Jan 18, 2021 @ 20:33:29.186';
+      login();
+      visitWithTimeRange(ALERTS_URL);
+      createTimelineOptionsPopoverBottomBar();
+      goToEsqlTab();
```
diving into this function, I think we should remove the `should('be.visible')` [here](https://github.com/michaelolo24/kibana/blob/unskip-discover-state-tests/x-pack/test/security_solution_cypress/cypress/tasks/common.ts#L88) (Cypress [already checks](https://docs.cypress.io/guides/core-concepts/interacting-with-elements) for visibility)

PhilippeOberti replied:
also I'm wondering if we could remove the about the `delay 500`? We're not using that delay in the expandable flyout Cypress tests and I don't remember seeing a test fail because of that

michaelolo24 replied:
Will remove should('be.visible'), and I figure the delay was put there because of how long it can take for the elements in the timeline tabs to load, but if it is necessary, there are definitely better ways of doing it besides an explicit delay time. Will look into it

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/esql/discover_timeline_state_integration.cy.ts:
```diff
@@ -104,13 +100,12 @@ describe.skip(
+      });
       it('should be able create an empty timeline with default discover state', () => {
         addNameToTimelineAndSave('Timerange timeline');
         createNewTimeline();
```
diving into this function, we should be able to remove the `should('be.visible')` [here](https://github.com/michaelolo24/kibana/blob/unskip-discover-state-tests/x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts#L305) (Cypress [already checks](https://docs.cypress.io/guides/core-concepts/interacting-with-elements) for visibility)

PhilippeOberti replied:
same thing like 123, looking into the `addDiscoverEsqlQuery` function, `should('be.visible')` from [here](https://github.com/michaelolo24/kibana/blob/unskip-discover-state-tests/x-pack/test/security_solution_cypress/cypress/tasks/discover.ts#L53)

PhilippeOberti replied:
also, line 136, looking into the `openTimelineFromSettings` function, I think we should be able to delete [this line](https://github.com/michaelolo24/kibana/blob/unskip-discover-state-tests/x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts#L388) entirely

PhilippeOberti replied:
let's remove it from [here](https://github.com/michaelolo24/kibana/blob/unskip-discover-state-tests/x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts#L408) as well, and also the `force: true`?

michaelolo24 replied:
Just gonna remove all of them in the timeline related files

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts:
```diff
@@ -359,6 +359,16 @@ export const saveTimeline = () => {
   });
 };
 
+export const confirmLeaveUnsavedTimeline = () => {
+  cy.get(DELETION_CONFIRMATION).click();
+};
+
+export const deleteTimeline = () => {
+  cy.get(TIMELINE_COLLAPSED_ITEMS_BTN).click();
+  cy.get(DELETE_TIMELINE_BTN).click();
+  cy.get(DELETION_CONFIRMATION).click();
+};
+
```
I'm getting errors where these constants are not defined

michaelolo24 replied:
Rebase error, will address, thanks!

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/esql/discover_timeline_state_integration.cy.ts:
```diff
@@ -212,6 +203,7 @@ describe.skip(
         navigateFromKibanaCollapsibleTo(STACK_MANAGEMENT_PAGE);
+        confirmLeaveUnsavedTimeline();
         cy.get(LOADING_INDICATOR).should('not.exist');
         goToSavedObjectSettings();
```
there is a chance we could remove line 11 and 12 of this [goToSavedObjectSettings](https://github.com/michaelolo24/kibana/blob/unskip-discover-state-tests/x-pack/test/security_solution_cypress/cypress/tasks/stack_management.ts#L10) no?

michaelolo24 replied:
Yep, will do

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/esql/discover_timeline_state_integration.cy.ts:
```diff
@@ -212,6 +203,7 @@ describe.skip(
+        confirmLeaveUnsavedTimeline();
         cy.get(LOADING_INDICATOR).should('not.exist');
         goToSavedObjectSettings();
         cy.get(LOADING_INDICATOR).should('not.exist');
```
should we extract the line below into a task function? (`cy.get(SAVED_OBJECTS_TAGS_FILTER).trigger('click');`)

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/esql/discover_timeline_state_integration.cy.ts:
what do you think about removing the nested describes (line 105 ` describe('save/restore', () => {` and line 194 `describe('saved search', () => {`)?
With clear text title I don't think the describe are necessary. They both have the same `beforeEach` code so that should work. It would make the file a bit easier to read.

Also I wonder if we could move the `const handleIntercepts` method above the top describe.

michaelolo24 replied:
Yea, will move the handle intercepts up, but I lean leaving the nested describe blocks as while all the `beforeEach` logic is the same, the context of what each block is testing is different, but I'll rename them to make them clearer

PhilippeOberti replied:
I'm not a fan of using `describe` just to separate tests like this but I'm not totally against it either 😄 

If we keep these nested `describe`, let's update the test titles to not contains the same string then? We probably don't need the repetition

michaelolo24 replied:
Yep, I had updated the titles in the last commit :)

PhilippeOberti reviewed (commented):
I left a few comments of extra things to potentially cleanup

michaelolo24 committed:
reintroduce task

michaelolo24 committed:
fix rebase errors and pr feedback cleanup

janmonschke committed:
Merge branch 'main' into pr/173308

janmonschke committed:
Merge branch 'main' into pr/173308

janmonschke committed:
Remove description test and check for titles

- The test title only mentions to check for title changes
- Description changes seem to be buggy in general and are not important at this point.

janmonschke committed:
Make sure we're not creating a saved search while a previous creation call is in progress

janmonschke committed:
the saved search update sometimes takes longer, so let's wait

janmonschke commented:
@kqualters-elastic Could you have a look at https://github.com/elastic/kibana/pull/173308/commits/e770e358b5e012ad2249b68315fef48f551fabee? I noticed that with the original fix, it would still sometimes create two saved searches.

lgestc commented on x-pack/plugins/security_solution/public/common/components/discover_in_timeline/use_discover_in_timeline_actions.tsx:
```diff
@@ -205,6 +205,10 @@ export const useDiscoverInTimelineActions = (
         // If no saved search exists. Create a new saved search instance and associate it with the timeline.
         try {
+          // Make sure we're not creating a saved search while a previous creation call is in progress
+          if (status !== 'idle') {
```
isnt there some kind of enum we could use here?

janmonschke replied:
@lgestc `status` comes from `@tanstack/query-core` and is union of string literals:

```ts
export type MutationStatus = 'idle' | 'loading' | 'success' | 'error'
```

So the type is checked, just not as an enum :)

lgestc reviewed (commented):
just one nit about magic value used

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.12|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 173308
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
